modules/widget-visibility/widget-conditions.php

```diff
@@ -199,6 +213,58 @@ public static function widget_admin_setup() {
 		wp_localize_script( 'widget-conditions', 'widget_conditions_parent_pages', $all_parents );
 	}
 
+	/**
+ 	 * Retrieves a full list of all pages, containing just the IDs, post_parent, and post_title fields.
+ 	 *
+ 	 * Since the WordPress' `get_pages` function does not allow us to fetch only the fields mentioned
+ 	 * above, we need to introduce a custom method using a direct SQL query fetching those.
+ 	 *
+ 	 * By fetching only those 3 fields and not populating the object cache for all the pages, we can 
+ 	 * improve the performance of the query on sites having a lot of pages.
+ 	 *
+ 	 * @see https://core.trac.wordpress.org/ticket/51469
+ 	 *
+ 	 * @return array List of all pages on the site (stdClass objects containing ID, post_title, and post_parent only).
+ 	 */
+	public static function get_pages() {
+		global $wpdb;
+
+		$last_changed = wp_cache_get_last_changed( 'posts' );
+		$cache_key    = "get_pages:$last_changed";
+		$pages        = wp_cache_get( $cache_key, 'widget_conditions' );
+		if ( false === $pages ) {
+			$pages = $wpdb->get_results( "SELECT {$wpdb->posts}.ID, {$wpdb->posts}.post_parent, {$wpdb->posts}.post_title FROM {$wpdb->posts} WHERE {$wpdb->posts}.post_type = 'page' AND {$wpdb->posts}.post_status = 'publish' ORDER BY {$wpdb->posts}.post_title ASC" );
+			wp_cache_set( $cache_key, $pages, 'widget_conditions' );
+		}
+
+		// Copy-pasted from the get_pages function. For usage in the `widget_conditions_get_pages` filter.
+		$parsed_args = array(
+			'child_of'     => 0,
+			'sort_order'   => 'ASC',
+			'sort_column'  => 'post_title',
+			'hierarchical' => 1,
+			'exclude'      => array(),
+			'include'      => array(),
+			'meta_key'     => '',
+			'meta_value'   => '',
+			'authors'      => '',
+			'parent'       => -1,
+			'exclude_tree' => array(),
+			'number'       => '',
+			'offset'       => 0,
+			'post_type'    => 'page',
+			'post_status'  => 'publish',
+		);
+
+		/**
+		 * Filters the retrieved list of pages.
+		 *
```

Comment:
```suggestion
		 *
		 * @module widget-visibility
		 *
		 * @since 9.1.0
		 *
```

We use to categorize hooks by feature in the Jetpack Codex.